### PR TITLE
feat(service-course-crawler): add course crawler

### LINKED ISSUE #28: Add course crawler
**Description**

### PR DESCRIPTION
**Pull request by ShirleyNekoDev**
**Fixes: #28**

**Checklist**

- [x] Commit messages follow conventionalcommits.org
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Changes**

- small fixes in common and news-crawler
- added initial course-crawler for crawling the current semester's courses

### CODE REVIEW

**Comment by ShirleyNekoDev** [comment]
Todo:
- [ ] write to database
- [ ] crawling TeleTask links
- [ ] archive crawling

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/CourseSeries.kt [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        EXERCISE,
+        BLOCK_SEMINAR;
+
+        override fun toString(): String {
+            return name.toLowerCase()
+        }
+
+        companion object {
+            fun parse(string: String) = string.toLowerCase()
+                .split(" / ")
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/CourseSeries.kt]: Spaces should be removed to catch more erroneous input as the parts are trimmed anyway.

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/CourseSeries.kt [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            Builder().apply(block).build()
+    }
+
+    class Builder {
+        lateinit var courseSeriesEntityLanguage: String
+        lateinit var title: String
+        lateinit var shortTitle: String
+        lateinit var abbreviation: String
+        var ects: Int = -1
+        var hoursPerWeek: Int = -1
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/CourseSeries.kt]: Why `-1`?

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/HpiCourseDetailPageCrawler.kt [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                    semester = Semester.parse(
+                        titleMatch.groups[3]!!.value,
+                        titleMatch.groups[2]!!.value
+                    )
+
+                    val laa = readLecturerAndAssistants(containerHtml)
+                    lecturers = laa.lecturers
+                    assistants = laa.assistants
+
+                    // TODO: search for teletask links
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/HpiCourseDetailPageCrawler.kt]: Should be moved to a new issue or just deleted

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/HpiCourseDetailPageCrawler.kt [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                                }
+                            }
+                        }
+                }
+            }
+        }
+    }
+
+    private fun parseDeadline(string: String): LocalDate? {
+        if(string.isBlank())
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/HpiCourseDetailPageCrawler.kt]: This line doesn't seem to be properly formatted

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/HpiCourseDetailPageCrawler.kt [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            try {
+                return LocalDate.parse(
+                    string.split(" ")
+                        .subList(0, 3)
+                        .joinToString(separator = " "),
+                    LONG_GERMAN_DATE_FORMAT
+                )
+            } catch (ex: IndexOutOfBoundsException) {
+            } catch (ex: DateTimeParseException) {
+            }
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/HpiCourseDetailPageCrawler.kt]: Exceptions shouldn't be silently ignored - at least add a comment explaining the reason

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import com.couchbase.client.java.view.ViewQuery
+import de.hpi.cloud.common.utils.couchbase.VIEW_BY_ID
+import de.hpi.cloud.common.utils.couchbase.withBucket
+import org.jsoup.Connection
+import org.jsoup.Jsoup
+import java.net.URI
+import java.net.URL
+import java.util.concurrent.atomic.AtomicInteger
+
+const val NAME = "HPI-MobileDev-Crawler[Courses]"
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt]: Should be the singular "course"

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+val CURRENT_SEMESTER = Semester(2019, Semester.Term.WINTER)
+val CRAWLERS = setOf(
+    HpiCourseListCrawler("IT-Systems Engineering", Degree.BACHELOR),
+    HpiCourseListCrawler("IT-Systems Engineering", Degree.MASTER),
+    HpiCourseListCrawler("Digital Health", Degree.MASTER),
+    HpiCourseListCrawler("Data Engineering", Degree.MASTER),
+    HpiCourseListCrawler("Cybersecurity", Degree.MASTER)
+)
+
+val requestCount = AtomicInteger(0)
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt]: Can be a simple Int

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    println()
+    bucket.query(ViewQuery.from("courseSeries", VIEW_BY_ID))
+        .allRows()
+        .forEach {
+            println("Remove CourseSeries ${it.id()}")
+            try {
+                bucket.remove(it.id())
+            } catch (ex: Exception) {
+            }
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt]: Don't silently ignore exceptions, at least add a comment

File: service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+fun deleteOldData(bucket: Bucket) {
+    println("Entered database cleanup mode\n")
+    bucket.query(ViewQuery.from("course", VIEW_BY_ID))
+        .allRows()
+        .forEach {
+            println("Remove Course ${it.id()}")
+            try {
+                bucket.remove(it.id())
+            } catch (ex: Exception) {
+            }
```
Comment (JonasWanke) [inline service-course-crawler/src/main/kotlin/de/hpi/cloud/course/crawler/Main.kt]: Don't silently ignore exceptions, at least add a comment

**Review by JonasWanke — CHANGES_REQUESTED** [review]

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**